File: _Archive/Overlap_Naive.java
package cs466_project;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.*;

public class Overlap_Naive {
	class Edge{
		String target;
		int weight;
		public Edge(String target, int weight) {
			this.target = target;
			this.weight = weight;
		}
		
		public String toString() {
			return target + '(' + weight + ')';
		}
	}

	public Map<String, List<Edge>> getEdge(String[] keys) {
		Map<String, List<Edge>> res = new HashMap<>();
		for (int i = 0; i < keys.length; i++) {
			String source = keys[i];
			res.put(source, new ArrayList<>());
			for (int j = 0; j < keys.length; j++) {
				if (i == j) continue;
				String target = keys[j];
				int weight = checkMatch(source, target);
				if (weight >= 1) {
					res.get(source).add(new Edge(target, weight));
				}
			}
		}
		return res;
	}
	int max = 0;
	String start = "";
	List<Edge> resultEdge = new ArrayList<>();
	
	public String getString(Map<String, List<Edge>> edges) {
		Map<String, List<Edge>> all = new HashMap<>();
		for (Object s : edges.keySet()) {
			Set<String> visited = new HashSet<>();
			List<Edge> res = new ArrayList<>();
			visited.add((String)s);
			dfs((String)s, visited, edges, res, (String)s);
			all.put((String)s, res);
		}
		
//		System.out.println(start);
//		System.out.println(edge);
//		System.out.println();
		StringBuilder sb = new StringBuilder(start);
		for (Edge e : resultEdge) {
			sb.append(e.target.substring(e.weight));
		}
//		System.out.println(max);
		return sb.toString();
		
	}
	
	private void dfs(String s, Set<String> visited, Map<String, List<Edge>> edges, List<Edge> res, String first) {
		if (visited.size() == edges.size()) {
			int cnt = 0;
			for (Edge e : res) {
				cnt += e.weight;
			}
			
			if (cnt > max) {
				start = (String)first;
				resultEdge = new ArrayList<>(res);
				max = cnt;
			}
			return;
		}
		List<Edge> edge = edges.get(s);
		for (Edge e : edge) {
			if (visited.contains(e.target)) continue;
			res.add(e);
			String next = e.target;
			visited.add(next);
			dfs(next, visited, edges, res, first);
			visited.remove(next);
			res.remove(e);
		}
	}
	
	private int checkMatch(String source, String target) {
		int sLen = source.length(), tLen = target.length();
		int res = source.length() - 1;
		while (!source.substring(sLen - res).equals(target.substring(0, res)) && res >= 0) {
			res--;
		}
		return res;
	}
	

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		String[] keys = {"CGAT", "ATCG", "GCAG", "AGCG", "GAGC", "GATC", "CAGC", "GGAG", "AGCA", "GCGA", "TCGG"};
		List<String> tmp = new ArrayList<>();
		try {
			String pathname = "/Users/tengdai/Documents/CS_Courses/Spring2019/CS466/Project/CS466_Project/rosalind_ba3c_copy_20.txt";
			File filename = new File(pathname);
//				InputStreamReader reader = new InputStreamReader(new FileInputStream(filename));
			BufferedReader br = new BufferedReader(new FileReader(filename));
			String line = "";
			line = br.readLine();
			while (line != null) {
				tmp.add(line);
				line = br.readLine();
			}
			br.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
//		String[] keys = tmp.toArray(new String[0]);
		Overlap_Naive solution = new Overlap_Naive();
		long start = System.currentTimeMillis();
		Map res = solution.getEdge(keys);
		String result = solution.getString(res);
		long finish = System.currentTimeMillis();
		long timeElapsed = finish - start;
		System.out.println(result);
		System.out.println("Running takes: " + (timeElapsed * 0.001) + " secondes.");

	}

}

File: _Archive/overlap_naive.py
import time

class OG_naive:
    def __init__(self):
        self.res_max = -1
        self.res_start = ''
        self.res_edges = []

    class Edge:
        def __init__(self, target, weight):
            self.target = target
            self.weight = weight
        def __str__(self):
            return str(self.target) + '(' + str(self.weight) + ')'



    def get_edge(self, keys):
        '''
        :param keys: input string list
        :return: dictionary, adjacent list of each input string
        '''
        res_dict = {}
        for i in range(len(keys)):
            source = keys[i]
            res_dict[source] = []
            for j in range(len(keys)):
                if i == j:
                    continue
                target = keys[j]
                weight = self.check_match(source, target)
                if (weight >= 2):
                    res_dict[source].append(self.Edge(target, weight))

        return res_dict

    def check_match(self, source, target):
        '''
        :param source: source string
        :param target: target string
        :return: overlap length
        '''
        s_len = len(source)
        res = s_len - 1
        while (res >= 0 and source[s_len - res:] != target[:res]):
            res -= 1
        return res

    def get_string(self, edges):
        '''
        :param edges: dictionary, adjacent list of each input string
        :return: shortest superstring contains all input string
        '''
        all_dict = {}
        for s in edges.keys():
            visited = set()
            res = []
            visited.add(s)
            self.dfs(s, visited, edges, res, s)
            all_dict[s] = res
        sb = self.res_start
        for e in self.res_edges:
            sb += e.target[e.weight:]
        return sb


    def dfs(self, s, visited, edges, res, first):
        '''
        :param s: current string
        :param visited: all visited string
        :param edges: adjacent edges map
        :param res: path list
        :param first: start string
        :return: nothing
        '''
        if (len(visited) == len(edges)):
            cnt = 0
            for e in res:
                cnt += e.weight
            if cnt > self.res_max:
                self.res_start = first
                self.res_edges = res.copy()
                self.res_max = cnt
            return

        edge = edges[s]
        for e in edge:
            if e.target in visited:
                continue
            res.append(e)
            next = e.target
            visited.add(next)
            self.dfs(next, visited, edges, res, first)
            visited.remove(next)
            res.remove(e)

# keys = ["CGAT", "ATCG", "GCAG", "AGCG", "GAGC", "GATC", "CAGC", "GGAG", "AGCA", "GCGA", "TCGG"]

file = open('rosalind_ba3c.txt', 'r')

keys = []

for row in file:

    keys.append(row.rstrip('\n'))

OG = OG_naive()
start = time.time()
edges = OG.get_edge(keys)
result = OG.get_string(edges)
end = time.time()
print('time', end - start) # sec: 119
print(result)
